File: sdk/typescript/examples/108-plan-execute-refs.ts
// Copyright (c) 2025 Agentspan
// Licensed under the MIT License.

/**
 * 108 — Plan-Execute with cross-step output piping via `Ref`.
 *
 * The `new Ref("step_id")` helper wires the **whole output** of an
 * upstream step into a downstream step's args. No JSON path, no field
 * selection, no internal task-ref naming to memorise — one expression
 * and the runtime substitutes the value at execution time.
 *
 * The example runs a three-step pipeline:
 *
 *     produce → enrich → report
 *
 * `produce` emits a record dict, `enrich` adds a derived field via
 * `Ref("produce")`, and `report` reads `Ref("enrich")` to format a
 * final summary. The plan is fully deterministic — no planner LLM
 * required — because we pass `plan` directly to `runtime.run`.
 *
 * Requirements:
 *   - Agentspan server running on http://localhost:6767 (or
 *     AGENTSPAN_SERVER_URL)
 *   - AGENTSPAN_LLM_MODEL set (default: openai/gpt-4o-mini)
 *
 * Run: npx tsx examples/108-plan-execute-refs.ts
 */

import {
  Agent,
  AgentRuntime,
  Op,
  Plan,
  Ref,
  Step,
  tool,
} from "../src/index.js";

const MODEL = process.env.AGENTSPAN_LLM_MODEL ?? "openai/gpt-4o-mini";

const produce = tool(
  async ({ record_id }: { record_id: string }) => ({
    record_id,
    value: 42,
    tags: ["alpha", "beta"],
  }),
  {
    name: "produce",
    description: "Return a fixed payload.",
    inputSchema: {
      type: "object",
      properties: { record_id: { type: "string" } },
      required: ["record_id"],
    },
  },
);

const enrich = tool(
  async ({ record }: { record: Record<string, unknown> }) => ({
    ...record,
    value_squared: ((record.value as number) ?? 0) ** 2,
  }),
  {
    name: "enrich",
    description: "Append a derived field. Reads the whole `produce` output via Ref.",
    inputSchema: {
      type: "object",
      properties: { record: { type: "object" } },
      required: ["record"],
    },
  },
);

const report = tool(
  async ({
    record,
    enriched,
  }: {
    record: { record_id: string; value: number; tags: string[] };
    enriched: { value_squared: number };
  }) => ({
    id: record.record_id,
    original_value: record.value,
    squared: enriched.value_squared,
    tags_joined: record.tags.join(", "),
    summary: `record=${record.record_id} value=${record.value} squared=${enriched.value_squared} tags=${JSON.stringify(record.tags)}`,
  }),
  {
    name: "report",
    description: "Format the final report. Reads BOTH upstream steps via Refs.",
    inputSchema: {
      type: "object",
      properties: {
        record: { type: "object" },
        enriched: { type: "object" },
      },
      required: ["record", "enriched"],
    },
  },
);

async function main() {
  const planner = new Agent({
    name: "ref_demo_planner",
    model: MODEL,
    instructions: "(planner unused; static plan supplied)",
  });

  const harness = new Agent({
    name: "ref_demo",
    model: MODEL,
    strategy: "plan_execute",
    planner,
    tools: [produce, enrich, report],
  });

  // Typed plan — no JSON strings, no field selectors. Each Ref serialises
  // to {"$ref": "<step_id>"} which the server rewrites to the right
  // Conductor template at compile time.
  const plan = new Plan({
    steps: [
      new Step("produce", {
        operations: [new Op("produce", { args: { record_id: "r-001" } })],
      }),
      new Step("enrich", {
        dependsOn: ["produce"],
        operations: [new Op("enrich", { args: { record: new Ref("produce") } })],
      }),
      new Step("report", {
        dependsOn: ["produce", "enrich"],
        operations: [
          new Op("report", {
            args: {
              record: new Ref("produce"),
              enriched: new Ref("enrich"),
            },
          }),
        ],
      }),
    ],
  });

  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(harness, "demo", { plan, timeoutSeconds: 120 });
    console.log(`status=${result.status} executionId=${result.executionId}`);
    await showPipelineOutputs(result.executionId);
  } finally {
    await runtime.shutdown();
  }
}

async function showPipelineOutputs(executionId: string) {
  const base = (process.env.AGENTSPAN_SERVER_URL ?? "http://localhost:6767/api")
    .replace(/\/api$/, "")
    .replace(/\/$/, "");
  const parent = (await (await fetch(`${base}/api/workflow/${executionId}?includeTasks=true`)).json()) as {
    tasks?: Array<{ referenceTaskName?: string; outputData?: { subWorkflowId?: string } }>;
  };
  let subId: string | undefined;
  for (const t of parent.tasks ?? []) {
    if (t.referenceTaskName?.endsWith("_plan_exec")) {
      subId = t.outputData?.subWorkflowId;
      break;
    }
  }
  if (!subId) return;
  const sub = (await (await fetch(`${base}/api/workflow/${subId}?includeTasks=true`)).json()) as {
    tasks?: Array<{ taskDefName?: string; outputData?: unknown }>;
  };
  console.log("\n── pipeline trace (Ref data flow) ────────────────────────");
  for (const t of sub.tasks ?? []) {
    if (["produce", "enrich", "report"].includes(t.taskDefName ?? "")) {
      console.log(`\n${t.taskDefName}:`);
      console.log(JSON.stringify(t.outputData, null, 2));
    }
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: sdk/python/examples/108_plan_execute_refs.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License.

"""108 — Plan-Execute with cross-step output piping via ``Ref``.

The ``Ref("step_id")`` helper wires the **whole output** of an upstream
step into a downstream step's args. No JSON path, no field selection,
no internal task-ref naming to memorise — one line of Python and the
runtime substitutes the value at execution time.

The pattern this enables:

    Step("fetch", operations=[Op("fetch_data", args={"url": URL})])
    Step("summarize", depends_on=["fetch"], operations=[
        Op("summarize", args={"document": Ref("fetch")}),
    ])

This example runs a three-step pipeline:

    produce → enrich → report

``produce`` emits a record dict, ``enrich`` adds a derived field via
``Ref("produce")``, and ``report`` reads ``Ref("enrich")`` to format a
final summary. The plan is fully deterministic — no planner LLM
required — because we pass ``plan=`` directly to ``runtime.run``.

What to look for in the output:
  * ``enrich`` receives the whole ``produce`` dict, not the literal
    ``{"$ref": "produce"}`` marker.
  * ``report`` reads ``enrich``'s output and ``produce``'s output
    independently (two Refs in the same args map).

Requirements:
  - AGENTSPAN_SERVER_URL=http://localhost:6767/api (default)
  - AGENTSPAN_LLM_MODEL=openai/gpt-4o-mini (default)
"""

from __future__ import annotations

import os

from agentspan.agents import AgentRuntime, Op, Plan, Ref, Step, plan_execute, tool


@tool
def produce(record_id: str) -> dict:
    """Emit a structured record. Step A."""
    return {
        "record_id": record_id,
        "value": 42,
        "tags": ["alpha", "beta"],
    }


@tool
def enrich(record: dict) -> dict:
    """Append a derived field. Step B reads Step A via ``Ref('produce')``."""
    return {
        **record,
        "value_squared": record["value"] ** 2,
    }


@tool
def report(record: dict, enriched: dict) -> dict:
    """Format the final report. Step C reads BOTH upstream steps."""
    return {
        "id": record["record_id"],
        "original_value": record["value"],
        "squared": enriched["value_squared"],
        "tags_joined": ", ".join(record["tags"]),
        "summary": (
            f"record={record['record_id']} value={record['value']} "
            f"squared={enriched['value_squared']} tags={record['tags']}"
        ),
    }


def main() -> None:
    harness = plan_execute(
        name="ref_demo",
        tools=[produce, enrich, report],
        planner_instructions="(planner unused; static plan supplied)",
        model=os.environ.get("AGENTSPAN_LLM_MODEL", "openai/gpt-4o-mini"),
    )

    plan = Plan(
        steps=[
            Step("produce", operations=[Op("produce", args={"record_id": "r-001"})]),
            Step(
                "enrich",
                depends_on=["produce"],
                operations=[Op("enrich", args={"record": Ref("produce")})],
            ),
            Step(
                "report",
                depends_on=["produce", "enrich"],
                operations=[
                    Op(
                        "report",
                        args={
                            "record": Ref("produce"),
                            "enriched": Ref("enrich"),
                        },
                    ),
                ],
            ),
        ],
    )

    with AgentRuntime() as runtime:
        result = runtime.run(harness, "demo", plan=plan, timeout=120)
        result.print_result()

        # The harness's final outputParameters don't surface per-step worker
        # results by default — print them explicitly so this example doubles
        # as a proof that `Ref()` actually carried the upstream dicts.
        _show_pipeline_outputs(result.execution_id)


def _show_pipeline_outputs(execution_id: str) -> None:
    """Walk into the plan_exec sub-workflow and dump the three step outputs."""
    import json

    import requests

    base = os.environ.get("AGENTSPAN_SERVER_URL", "http://localhost:6767/api")
    base_url = base.rstrip("/").replace("/api", "")

    parent = requests.get(
        f"{base_url}/api/workflow/{execution_id}?includeTasks=true", timeout=10
    ).json()
    sub_id = None
    for t in parent.get("tasks", []):
        if t.get("referenceTaskName", "").endswith("_plan_exec"):
            sub_id = (t.get("outputData") or {}).get("subWorkflowId")
            break
    if not sub_id:
        return

    sub = requests.get(
        f"{base_url}/api/workflow/{sub_id}?includeTasks=true", timeout=10
    ).json()
    print("\n── pipeline trace (Ref data flow) ────────────────────────")
    for t in sub.get("tasks", []):
        name = t.get("taskDefName")
        if name in ("produce", "enrich", "report"):
            print(f"\n{name}:")
            print(json.dumps(t.get("outputData", {}), indent=2))


if __name__ == "__main__":
    main()
